cmd/gitserver/internal/git/gitcli/commitlog.go

```diff
@@ -0,0 +1,196 @@
+	if opt.AllRefs {
+		args = append(args, "--all")
+	}
+
+	if opt.IncludeModifiedFiles {
+		args = append(args, "--name-only")
+	}
+	if opt.FollowPathRenames {
+		args = append(args, "--follow")
+	}
+	if opt.Path != "" {
+		args = append(args, "--", opt.Path)
+	}
+
+	return args, nil
+}
+
+func newCommitLogIterator(repoName api.RepoName, spec string, r io.ReadCloser) *commitLogIterator {
+	commitScanner := bufio.NewScanner(r)
+	// We use an increased buffer size since sub-repo permissions
+	// can result in very lengthy output.
+	commitScanner.Buffer(make([]byte, 0, 65536), 4294967296)
+	commitScanner.Split(commitSplitFunc)
+
+	return &commitLogIterator{
+		Closer:   r,
+		repoName: repoName,
+		spec:     spec,
+		sc:       commitScanner,
+	}
+}
+
+type commitLogIterator struct {
+	io.Closer
+	repoName api.RepoName
+	spec     string
+	sc       *bufio.Scanner
+}
+
+func (it *commitLogIterator) Next() (*git.GitCommitWithFiles, error) {
+	if !it.sc.Scan() {
+		if err := it.sc.Err(); err != nil {
+			// If exit code is 128 and `fatal: bad object` is part of stderr, most likely we
+			// are referencing a commit that does not exist.
+			// We want to return a gitdomain.RevisionNotFoundError in that case.
+			var e *CommandFailedError
+			if errors.As(err, &e) && e.ExitStatus == 128 {
+				if bytes.Contains(e.Stderr, []byte("not a tree object")) || bytes.Contains(e.Stderr, []byte("fatal: bad object")) {
+					return nil, &gitdomain.RevisionNotFoundError{Repo: it.repoName, Spec: it.spec}
+				}
+				if bytes.Contains(e.Stderr, []byte("fatal: ambiguous argument")) && bytes.Contains(e.Stderr, []byte("unknown revision or path not in the working tree.")) {
+					return nil, &gitdomain.RevisionNotFoundError{Repo: it.repoName, Spec: it.spec}
+				}
+				if bytes.Contains(e.Stderr, []byte("fatal: your current branch")) && bytes.Contains(e.Stderr, []byte("does not have any commits yet")) {
+					return nil, io.EOF
+				}
+			}
+			return nil, err
+		}
+		return nil, io.EOF
+	}
+
+	rawCommit := it.sc.Bytes()
+	commit, err := parseCommitFromLog(rawCommit)
+	if err != nil {
+		return nil, err
+	}
+	return commit, nil
+}
+
+func (it *commitLogIterator) Close() error {
+	err := it.Closer.Close()
+	if err != nil {
+		var e *CommandFailedError
+		if errors.As(err, &e) && e.ExitStatus == 128 {
+			if bytes.Contains(e.Stderr, []byte("not a tree object")) || bytes.Contains(e.Stderr, []byte("fatal: bad object")) {
+				return &gitdomain.RevisionNotFoundError{Repo: it.repoName, Spec: it.spec}
+			}
+			if bytes.Contains(e.Stderr, []byte("fatal: your current branch")) && bytes.Contains(e.Stderr, []byte("does not have any commits yet")) {
+				return nil
+			}
+		}
+	}
+	return err
+}
+
+func commitSplitFunc(data []byte, atEOF bool) (advance int, token []byte, err error) {
+	if len(data) == 0 {
+		// Request more data
+		return 0, nil, nil
+	}
+
+	// Safety check: ensure we are always starting with a record separator
+	if data[0] != '\x1e' {
```

Comment:
nit: can we extract that into a constant?

```suggestion
const recordSeparator = '\x1e'

if data[0] != recordSeparator {
```